### PR TITLE
Enable channels_last_3d on SyncBatchNorm

### LINKED ISSUE #88021: SyncBatchNorm 3D is slow with torch.channels_last_3d format in DDP on GPU
### 🐛 Describe the bug

SyncBatchNorm runs slower with torch.channels_last_3d memory format ( DistributedDataParallel  multi-gpu training), even though channels last format is supposed to improve the performance on GPUs. 

Upon closer inspection of SyncBatchNorm:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/_functions.py#L10

https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/_functions.py#L107 

it seems we only check for torch.channels_last (2D case), and if the data is in torch.channels_last_3d (3D) it always gets converted to contiguous(). Commenting those lines improves the speed. 

the channels_last support  was introduced in https://github.com/pytorch/pytorch/pull/46906
the inline subfunction seems to return True only for  torch.channels_last (not torch.channels_last_3d)

```
inline bool batch_norm_use_channels_last_kernels(const at::Tensor& self) {
  return self.is_contiguous(at::MemoryFormat::ChannelsLast) || self.ndimension() == 2;
}
```

https://github.com/pytorch/pytorch/pull/46906/files#diff-e895801158bc7155ba3b61826cdd129c5e2caa5c3519171abfbe9bbe5406f52fR3

### Versions

PyTorch version: 1.13.0a0+d0d6b1f
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.22.2
Libc version: glibc-2.31

Python version: 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:10)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-122-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A40
GPU 1: NVIDIA A40
GPU 2: NVIDIA A40
GPU 3: NVIDIA A40
GPU 4: NVIDIA A40
GPU 5: NVIDIA A40
GPU 6: NVIDIA A40
GPU 7: NVIDIA A40

Nvidia driver version: 470.129.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] functorch==0.3.0a0
[pip3] numpy==1.22.2
[pip3] pytorch-quantization==2.1.2
[pip3] torch==1.13.0a0+d0d6b1f
[pip3] torch-tensorrt==1.3.0a0
[pip3] torchtext==0.11.0a0
[pip3] torchvision==0.14.0a0
[conda] functorch                 0.3.0a0                  pypi_0    pypi
[conda] mkl                       2020.4             h726a3e6_304    conda-forge
[conda] mkl-include               2020.4             h726a3e6_304    conda-forge
[conda] numpy                     1.22.2           py38h6ae9a64_0    conda-forge
[conda] pytorch-quantization      2.1.2                    pypi_0    pypi
[conda] torch                     1.13.0a0+d0d6b1f          pypi_0    pypi
[conda] torch-tensorrt            1.3.0a0                  pypi_0    pypi
[conda] torchtext                 0.11.0a0                 pypi_0    pypi
[conda] torchvision               0.14.0a0                 pypi_0    pypi

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### PR DESCRIPTION
**Pull request by xwang233**
This PR enabled the use of fast channels_last kernels on SyncBatchNorm with channels_last_3d memory format.

With a small benchmark script here https://github.com/pytorch/pytorch/issues/88021#issuecomment-1299059859, on V100, I got 

master:
```
DDP channels_last=False, run_forward_backward, time: 0.8945400714874268 sec
DDP channels_last=True, run_forward_backward, time: 1.4736433029174805 sec
```

This PR:
```
DDP channels_last=False, run_forward_backward, time: 0.8927242755889893 sec
DDP channels_last=True, run_forward_backward, time: 0.48697471618652344 sec
```

This PR is a follow-up of https://github.com/pytorch/pytorch/pull/46906

Close https://github.com/pytorch/pytorch/issues/88021

### CODE REVIEW

**Comment by myron** [comment]
@xwang233 Looks great, thank you so much for adding it!

**Comment by xwang233** [comment]
Hi @ngimel , would you take a look at this PR? All tests are passed. Thanks!

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -5319,6 +5319,54 @@ def test_post_localSGD_optimizer_step_reload(self):
...
+            bs_offset = int(rank * 2)
+            global_bs = int(num_processes * 2)
+
+            model = ONLY_SBN_NET
+            model_gpu = copy.deepcopy(model).cuda(rank)
+            model_DDP = nn.parallel.DistributedDataParallel(
+                model_gpu, device_ids=[rank]
+            )
+
+            memory_format = torch.channels_last_3d
```
Comment (ngimel) [inline torch/testing/_internal/distributed/distributed_test.py:5340]: looks like preparing inputs is the only difference with channels last 2d, can you refactor everything else into a helper function for these 2 tests?

**Review by ngimel — COMMENTED** [review]
Minor testing suggestion, otherwise looks good, thanks for the fix!

**Comment by xwang233** [comment]
Hi @ngimel , I've merged the test code with channels_last and channels_last_3d into one test. All the CI checks look green now.

**Review by ngimel — APPROVED** [review]
Great, thanks!

**Pull request closed without merging**